File: frontend_v2/src/features/financial/lib/ledger-timeline.ts
import { z } from "zod"
export const ledgerTimeline = z.object({
  case_id: z.string(),
  account_id: z.string().nullable(),
  start_date: z.string().nullable(),
  end_date: z.string().nullable(),
  population: z.enum(["working", "verified"]),
  snapshot_sha256: z.string().regex(/^[a-f0-9]{64}$/),
  excluded_rows: z.number().int().nonnegative(),
  limitation: z.string(),
  rows: z
    .array(
      z.object({
        key: z.string(),
        source_document_id: z.string(),
        account_id: z.string(),
        account_label: z.string(),
        chronology_date: z.string().regex(/^\d{4}-\d{2}-\d{2}$/),
        chronology_basis: z.enum([
          "value_date",
          "transaction_date",
          "posted_date",
          "effective_date",
          "ordering_date",
          "statement_end_ordering_only",
        ]),
        ordering_date: z.string(),
        direction: z.enum(["debit", "credit"]),
        amount_minor: z.string().regex(/^(0|[1-9][0-9]*)$/),
        currency: z.string(),
        proof_class: z.enum(["p0", "p1", "p2", "p3"]),
        description: z.string().nullable(),
      })
    )
    .max(1000),
})
export type LedgerTimeline = z.infer<typeof ledgerTimeline>
export const chronologyLabels = {
  value_date: "Value date",
  transaction_date: "Transaction date",
  posted_date: "Posted date",
  effective_date: "Effective date",
  ordering_date: "Ordering date only",
  statement_end_ordering_only:
    "Transaction date unknown — statement-end ordering only",
}

export function validContextDate(value: unknown): value is string {
  if (typeof value !== "string" || !/^\d{4}-\d{2}-\d{2}(?:$|T| )/.test(value))
    return false
  const day = value.slice(0, 10)
  const stamp = new Date(day + "T00:00:00Z")
  return (
    !Number.isNaN(stamp.getTime()) && stamp.toISOString().slice(0, 10) === day
  )
}

File: backend/services/financial/ledger_timeline.py
"""Chronology of captured ledger readings; dates never imply event correlation."""
import json
from services.financial.ledger_summary import LedgerSummaryError

MAX_TIMELINE_ROWS = 1000


def ledger_timeline(export, *, population="working"):
    if population not in ("working", "verified"):
        raise LedgerSummaryError("Choose working or verified readings.")
    document = json.loads(export.snapshot.content)
    ledger = document["ledger"]
    if not document.get("export_ready") or not ledger.get("history_captured"):
        raise LedgerSummaryError("Timeline requires a consistent captured ledger and history.")
    eligible = [r for r in ledger["readings"] if (
        r["exclusion_reason"] in (None, "proof_class_not_included")
        if population == "working" else r["included"]
    )]
    if len(eligible) > MAX_TIMELINE_ROWS:
        raise LedgerSummaryError("More than 1,000 current readings match. Narrow the account or ordering-date scope; no partial chronology was returned.")
    rows = []
    for reading in eligible:
        row = reading["row"]
        basis = next((field for field in ("value_date", "transaction_date", "posted_date", "effective_date") if row.get(field)), "ordering_date")
        if row.get("ordering_date_context") == "statement_end_ordering_only":
            basis = "statement_end_ordering_only"
        rows.append(dict(
            **row,
            chronology_date=row["ordering_date"] if basis == "statement_end_ordering_only" else row[basis],
            chronology_basis=basis,
            account_label=reading.get("account", {}).get("label") or "Account " + row["account_id"][:8],
        ))
    rows.sort(key=lambda r: (r["chronology_date"], r["key"]))
    return dict(
        case_id=ledger["case_id"], account_id=ledger["account_id"],
        start_date=ledger["start_date"], end_date=ledger["end_date"],
        population=population, snapshot_sha256=export.snapshot.sha256,
        rows=rows, excluded_rows=len(ledger["readings"])-len(eligible),
        limitation="Ledger scope uses ordering dates. Display uses value date, then transaction, posted, effective or ordering date, explicitly labelled. A displayed date can fall outside the ordering-date filter. Same-day position does not establish sequence. Case events are context, not matched or corroborated payments.",
    )
